### DISCUSSION TITLE
[BE201] Nginx server 錯誤

### DISCUSSION THREAD

**Discussion by powerfultraveling** [Lidemy/forum#186] [後端相關問題]
## 問題
error nginx: [emerg] dlopen() "/usr/local/nginx/modules/ds_http_module.so" failed

設定好 ufw 之後，使用 `sudo systemctl start nginx` 的指令，terminal 上面跑出了以下的訊息:
![擷取](https://user-images.githubusercontent.com/81896228/129830335-80d5a8a0-fff2-4419-925b-b6943c186208.PNG)

根據上面的資訊，我推斷 Nginx 有成功開啟運行，但是之後因為一些 configuration 的問題，所以沒辦法運行。

於是我搜尋了error nginx: [emerg] dlopen() "/usr/local/nginx/modules/ds_http_module.so" failed，找到了一些資料和文章，試過了一些辦法，但是眾說紛紜，沒辦法特定出 bug 是出在哪裡，所以想請教老師我該怎麼找相關的資料，或是有任何解決辦法嗎?

## 問題背景知識
這週試著在 AWS 的 ec2 上面把 node.js 留言板架設上去。一開始跟著老師課堂上的步驟做完之後，在瀏覽器上輸入網域之後有正確跑到 Nginx 的 landing page。

但是我的 app 是開在 port 3000，所以我先執行 app 並在網域後面加上 :3000，結果是沒辦法跑出來。後來用 localhost 來跑一次，發現沒有問題，所以排除了 app 本身有 bug 的疑慮。

後來看了很多文章，嘗試修改了 /etc/nginx/sites-available/default 的設定，看看有沒有辦法 debug，但後來跑出了更多不同的 error，因為過程有點混亂，最後決定刪掉 Nginx 重頭來。然後就到了現在的狀況。

**Comment by aszx87410** [comment]
我猜是你修改的  /etc/nginx/sites-available/default 設定導致的，就算刪掉 nginx 重來，也要看一下設定是不是你改過的版本
也可以試試看砍掉 nginx 也砍掉  /etc/nginx/sites-available/default 再重裝一次，看看是不是恢復正常
先從找到問題根源開始

**Reply by powerfultraveling** [reply]
# 上面的問題已解決，但仍舊無法連線到其他 port

## 問題解決過程
我在之前其實也已經砍過 nginx ，並且 cd 到 etc/nginx 確認 configuration 的檔案也已刪除。後來發現，自己似乎還是有一些東西沒刪乾淨，建議要在 ubuntu 裡面刪東西的可以使用 `whereis nginx` 的指令確認系統裡相關的資料都已經刪乾淨。

## 尚未解決的問題
如最前面所講的，一開始去改 config 檔是因為沒辦法透過在 domain name/ IP 位置後面加 port 碼來連線到放在那邊的 app，現在雖然解決了剛剛的 error，但是又跑回了最初的問題「無法連接上 default port 80 以外的 port。

#### 我這邊寫了一個叫做 test.js 的檔案，用簡單的 express server 當作測試:
![testjs](https://user-images.githubusercontent.com/81896228/129885273-9ad83278-717b-4ea1-bcfc-57dc15843232.PNG)

#### 確認 pm2 已經開啟，並且 test.js 正在被執行中:
![pm2 s](https://user-images.githubusercontent.com/81896228/129885594-ac86630b-05a6-4f43-930a-691104c31c11.PNG)

#### 確認 ufw 的設定已經打開 port 3001:
![ufw](https://user-images.githubusercontent.com/81896228/129885876-796d52ff-6c43-457e-a1ee-ca94867fcc61.PNG)

#### 確認 localhost 來跑的話，test.js 是正常運作的:
![loca](https://user-images.githubusercontent.com/81896228/129886159-b9fbe475-acdc-404d-971a-0e01b2584f8a.PNG)

全部確認之後在 browser 裡面輸入我的 domain name powerfultraveling.site，有成功連上 Nginx 的 landing page ，但是鍵入 powerfultraveling.site:3001 後仍舊沒辦法連上 port 3001。

## 補充
在最一開始我以為是config 檔導致沒辦法連上，但後來想了一下，config 檔應該是用於設定 Proxy，所以應該不會是在那邊解決目前的問題。

**Reply by aszx87410** [reply]
如果你是用 AWS 的話，你 AWS 那邊有開 3001 port 嗎
